### PR TITLE
Sleep State & Native Death Notices

### LINKED ISSUE #9: npc sleepstates do not work
witnessed in the office complex levels where in the freezer, the scientists are already dead due to the zombies sleepstates not correctly being set

### PR DESCRIPTION
**Pull request by ampreeT**
This pull request addresses a problem with AI sleep states and adds native death notices that appear on the HUD. `CAI_BaseNPC::UpdateSleepState` has been fully reconstructed with the singleplayer checks removed so AI should properly be able to sleep. Native death notices are now supported via fake clients.

Fixes #9

image

**Changes:**

- Added 1 signature and 3 vtable indexes to gamedata for bms and hl2dm
- Reconstructed `CAI_BaseNPC::UpdateSleepState` without singleplayer checks
- Added native death notices
    - Uses a maximum of 2 fake clients
    - Shows AI kills, AI suicides and player kills on AI

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,72 +1,74 @@
...
-- Enables players to play together on single-player campaigns or maps
-- Complements existing cooperative campaigns or maps
-- Fixes various crashes
-- Re-enables single-player functionality
-- Includes map configs adjusting for cooperative play
-- Comes with optional addons, enhancing your experience
-- Can be used by server owners as the ultimate map config tool
-- Is built on [SourceMod](https://www.sourcemod.net)
-- __Only needs to be installed on the server__.
+SourceCoop is a cooperative mod for Source Engine games enabling singleplayer campaigns to be played together. It currently supports [Black Mesa](https://store.steampowered.com/app/362890/Black_Mesa/) and [Half-Life 2: Deathmatch](https://store.steampowered.com/app/362890/Black_Mesa/).
```
Comment (Alienmario) [inline README.md:12]: and [Half-Life 2: Deathmatch](https://store.steampowered.com/app/362890/Black_Mesa/).

Comment (Alienmario) [reply README.md:12]: wrong link

Comment (ampreeT) [reply README.md]: Fixed the link for Half-life 2: Deathmatch.

File: scripting/include/srccoop/classdef.inc [hunk 2]
```diff
@@ -1850,31 +2398,80 @@ methodmap IServerGameDLL
...
 methodmap CTraceRay < Handle
 {
-	public CTraceRay(const float vecPosition[3], const float vecRayType[3], const int iFlags, const RayType rayType, const TraceEntityFilter traceFilter, const any pData)
+	public CTraceRay(const float vecPosition[3], const float vecRayType[3], const int iTraceMask, const RayType eRayType, const TraceEntityFilter fnFilter, const any pData)
 	{
-		return view_as<CTraceRay>(TR_TraceRayFilterEx(vecPosition, vecRayType, iFlags, rayType, traceFilter, pData));
+		return view_as<CTraceRay>(TR_TraceRayFilterEx(vecPosition, vecRayType, iTraceMask, eRayType, fnFilter, pData));
+	}
+
+	public Handle GetHandle()
```
Comment (Alienmario) [inline scripting/include/srccoop/classdef.inc:2406]: why is this needed?

File: scripting/include/srccoop/deathnotice.inc [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+			// If the death notification displays before the player name is
+			// updated, then the old name will be displayed which is unintended behavior.
+			float flDelay = 0.35; 
+
+			// Pushes a entry a new queue if a queue does not exist.
+			// If a queue does exist, pushes a entry to the queue. 
+			g_sDnData.m_sList.PushArray(sEntry, sizeof(sEntry));
+			if (g_sDnData.m_hTimer == INVALID_HANDLE)
+			{
+				g_sDnData.m_hTimer = CreateTimer(flDelay, DeathNotice_Timer, _, TIMER_REPEAT);
```
Comment (Alienmario) [inline scripting/include/srccoop/deathnotice.inc:413]: Not sure if mapchange handling is needed for the timer here

File: scripting/include/srccoop/classdef.inc [hunk 4]
```diff
@@ -848,6 +1041,163 @@ methodmap CSceneEntity < CBaseEntity
...
+	public void ActiveList(CSound pSound)
+	{
+		pSound.m_pSoundEnt = view_as<int>(this);
+		pSound.m_iSoundIndex = GetEntProp(this.GetEntIndex(), Prop_Data, "m_iActiveSound");
+	}
+}
+
+enum struct CSound
+{
+	int m_pSoundEnt;
```
Comment (Alienmario) [inline scripting/include/srccoop/classdef.inc:1069]: Is there a reason for not storing as CSoundEnt directly?

File: scripting/include/srccoop/classdef.inc [hunk 5]
```diff
@@ -742,6 +820,108 @@ methodmap CBaseEntity
...
+		pTraceRay.Close();
+		return false;
+	}
+	public bool FVisiblePosition(const float vec3TargetOrigin[3], int iTraceMask = MASK_BLOCKLOS, CBaseEntity& pBlocker = NULL_CBASEENTITY)
+	{
+		float vec3LookerOrigin[3];
+		this.GetEyePosition(vec3LookerOrigin);
+
+		CTraceRay pTraceRay;
+		if (FindConVar("ai_los_mode").BoolValue)	// TODO: Cache this ConVar.
```
Comment (Alienmario) [inline scripting/include/srccoop/classdef.inc:890]: can be static

File: scripting/include/srccoop/classdef.inc [hunk 6]
```diff
@@ -383,6 +440,27 @@ methodmap CBaseEntity
 	{
 		SetEntProp(this.GetEntIndex(), Prop_Data, "m_spawnflags", iSpawnFlags);
 	}
+	public void GetEyePosition(float vecBuffer[3])
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc:443]: I thought about leaving it in `CBaseCombatCharacter` but the game defines this function as a virtual `CBaseEntity::GetEyePosition` and it is used in `CBaseEntity::FVisible` directly without a player check. I know SourcePawn doesn't have the concept of traits like other language so that is why I am doing a "virtual" dispatch in `CBaseEntity`.

File: README.md [hunk 7]
```diff
@@ -1,72 +1,74 @@
...
-- Can be used by server owners as the ultimate map config tool
-- Is built on [SourceMod](https://www.sourcemod.net)
-- __Only needs to be installed on the server__.
+SourceCoop is a cooperative mod for Source Engine games enabling singleplayer campaigns to be played together. It currently supports [Black Mesa](https://store.steampowered.com/app/362890/Black_Mesa/) and [Half-Life 2: Deathmatch](https://store.steampowered.com/app/320/HalfLife_2_Deathmatch/).
+
+- Supports campaigns.
+- Restores singleplayer functionality.
+- Handles world, player and equipment persistence between maps.
+- Easy to use optional addons.
+- Easy to convert maps from singleplayer into cooperative.
```
Comment (Alienmario) [inline README.md:18]: well if u're a skilled mapper or done some scripting then yea

File: README.md [hunk 8]
```diff
@@ -1,72 +1,74 @@
...
-- Includes map configs adjusting for cooperative play
-- Comes with optional addons, enhancing your experience
-- Can be used by server owners as the ultimate map config tool
-- Is built on [SourceMod](https://www.sourcemod.net)
-- __Only needs to be installed on the server__.
+SourceCoop is a cooperative mod for Source Engine games enabling singleplayer campaigns to be played together. It currently supports [Black Mesa](https://store.steampowered.com/app/362890/Black_Mesa/) and [Half-Life 2: Deathmatch](https://store.steampowered.com/app/320/HalfLife_2_Deathmatch/).
+
+- Supports campaigns.
+- Restores singleplayer functionality.
+- Handles world, player and equipment persistence between maps.
```
Comment (Alienmario) [inline README.md:16]: Doesn't really handle world persistence. Maps have to be painstakingly edited using global states to fake that.

File: scripting/include/srccoop/deathnotice.inc [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+			}
+		}
+		else if (g_sDnData.m_piNoticeLevel.IntValue == 1)
+		{
+			if (sEntry.m_bSuicide)
+			{
+				PrintToChatAll(SRCCOOP_CHAT_TAG..."%s suicided!", sEntry.m_szTarget2Name);
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/deathnotice.inc]: Changed all instances of `PrintToChatAll` to `MsgAllNoSrv`.

File: scripting/include/srccoop/classdef.inc [hunk 10]
```diff
@@ -1086,10 +1444,90 @@ methodmap CAI_BaseNPC < CBaseCombatCharacter
...
+	}
 	public void SetSleepState(const int iSleepState)
 	{
 		SetEntProp(this.GetEntIndex(), Prop_Data, "m_SleepState", iSleepState);
 	}
+	public void Wake()
+	{
+		this.AcceptInputStr("Wake");
+	}
+	public void Sleep()
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc:1518]: Yep I just double checked this and it doesn't look like there is a `Sleep` input. We will have to reimplement this function later.

File: scripting/include/srccoop/classdef.inc [hunk 11]
```diff
@@ -256,6 +296,23 @@ methodmap CBaseEntity
 	{
 		return CBaseEntity(CreateEntityByName(szClassname));
 	}
+	//public static CBaseEntity FromAddress(const Address pAddress)
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Removed the function for now as it was unused. Maybe we should consider moving the version from utils into `CBaseEntity`.

File: scripting/include/srccoop/classdef.inc [hunk 12]
```diff
@@ -383,6 +440,27 @@ methodmap CBaseEntity
...
+		}
+		else
+		{
+			float vec3Offset[3];
+			GetEntPropVector(this.GetEntIndex(), Prop_Data, "m_vecViewOffset", vec3Offset);
+			
+			this.GetAbsOrigin(vecBuffer);
+			vecBuffer[0] += vec3Offset[0];
+			vecBuffer[1] += vec3Offset[1];
+			vecBuffer[2] += vec3Offset[2];
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Changed the implementation to use `AddVectors`.

File: scripting/include/srccoop/classdef.inc [hunk 13]
```diff
@@ -1701,6 +2159,96 @@ methodmap CParticleSystem < CBaseEntity
...
+		if (bOutOfBounds)
+			ThrowError("Received out of bounds index %d; range is 0-%i", MaxClients);
+		
+		int iConnectedOffset = GetEntSendPropOffs(this.GetEntIndex(), "m_bConnected");
+		if (iConnectedOffset == -1)
+			ThrowError("Unable to retrieve offset %s", "m_iTeam");
+		
+		Address pAddress = this.GetAddress();
+		Address pOffset = view_as<Address>(view_as<int>(pAddress) + iConnectedOffset + iEntIndex * 4);
+		return LoadFromAddress(pOffset, NumberType_Int32);
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Removed the explicit casts and changed instances of `LoadFromAddress` to use `Deref`.

File: scripting/include/srccoop/classdef.inc (hunk 4 above)
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: No there currently isn't any reason. Changed implementation to directly use `CSoundEnt`.

File: scripting/include/srccoop/classdef.inc (hunk 5 above)
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Madethe  `FindConVar` a global variable that gets initialized once. Fixed.

File: README.md (hunk 7 above)
Comment (ampreeT) [reply README.md]: Changed the description of that line a bit to reflect that

File: README.md (hunk 8 above)
Comment (ampreeT) [reply README.md]: Yeah I was thinking more of the lines around the plugin remembering the player backtracking but that doesn't fit. Changed the description by removing the part about world persistence.

File: scripting/include/srccoop/deathnotice.inc (hunk 3 above)
Comment (ampreeT) [reply scripting/include/srccoop/deathnotice.inc:427]: Added removal of timer data in for good practice. I don't think anyone will be joining that fast either way lol

File: scripting/include/srccoop/deathnotice.inc [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+		{
+			// Delay is used to allow time for the game to update the player name.
+			// If the death notification displays before the player name is
+			// updated, then the old name will be displayed which is unintended behavior.
+			float flDelay = 0.35; 
+
+			// Pushes a entry a new queue if a queue does not exist.
+			// If a queue does exist, pushes a entry to the queue. 
+			g_sDnData.m_sList.PushArray(sEntry, sizeof(sEntry));
+			if (g_sDnData.m_hTimer == INVALID_HANDLE)
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/deathnotice.inc]: Fixed that.

File: scripting/include/srccoop/classdef.inc (hunk 11 above)
[the remark this answers is not in the document]
Comment (Alienmario) [reply scripting/include/srccoop/classdef.inc]: I do prefer using classdef for simple accessors/calls and having complex functions somewhere else, like utils.
What could work is having a method delegate a call to utils, while wrapping wrapping the result in CBaseEntity.

File: scripting/include/srccoop/classdef.inc (hunk 2 above)
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc:2391]: Are you talking about `GetHandle`? I'm using it like this to limit the number of casts.

If you are talking about the other member functions, I need them for the implementation of FVisible.

Comment (Alienmario) [reply scripting/include/srccoop/classdef.inc:2391]: But if you inherit from Handle you should not need to cast anyway.

File: scripting/include/srccoop/deathnotice.inc [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+				g_sDnData.m_pBot2.SetTeam(TEAM_SPECTATOR);
+			}
+			return g_sDnData.m_pBot2;
+		}
+		return NULL_CBASEENTITY;
+	}
+	// Hook callback to hide player name changes for bots.
+	// Bots have to change their names in order to display a death display.
+	//
+	public static Action TextMsg(BfRead bf)
```
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/deathnotice.inc:160]: Just tried your snippet and I don't think this works in Black Mesa. I've tried using "player_changename" in the past but I've never seen the event fire.

File: scripting/include/srccoop/classdef.inc (hunk 11 above)
[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Implemented `FromAddress` back in as a wrapper around `GetEntityFromAddress`.

File: scripting/include/srccoop/deathnotice.inc (hunk 15 above)
[the remark this answers is not in the document]
Comment (Alienmario) [reply scripting/include/srccoop/deathnotice.inc:160]: In that case you can add a flag only set while changing the bot name and check for it, or hook/unhook TextMsg on demand.
Just tested it's called inside SetClientName.

File: scripting/include/srccoop/classdef.inc (hunk 2 above)
Comment (ampreeT) [reply scripting/include/srccoop/classdef.inc]: Oversight on my part. Fixed this by removing `GetHandle` and using `this` directly.
